### PR TITLE
Gml 1843 async pytg

### PR DESCRIPTION
**Pull request by luzhoutg**

### CODE REVIEW

**Review by qe-tigergraph — COMMENTED** [review]
Unit Test: SUCCESS, e2e Test: SKIPPED, Jenkins_job:http://192.168.99.101:30080/job/mlwb_build/1148/

**Review by qe-tigergraph — APPROVED** [review]
QE Approved

**Review by qe-tigergraph — COMMENTED** [review]
Unit Test: SUCCESS, e2e Test: SKIPPED, Jenkins_job:http://192.168.99.101:30080/job/mlwb_build/1149/

**Review by qe-tigergraph — APPROVED** [review]
QE Approved

**Comment by luzhoutg** [comment]
1. renamed the folder "async_ver2" to "pytgasync" and update all path
2. solved the conflicts with master
3. formatted code
4. added test case using "getVertexCount" and "getEdgeCount"
5. passed all unit tests

File: tmp/Cora/._ingest.gsql
Comment (parkererickson-tg) [inline tmp/Cora/._ingest.gsql:1]: The temp folder here should be removed

File: tests/testserver.json
Comment (parkererickson-tg) [inline tests/testserver.json:1]: does this typo actually work?

File: pyTigerGraph/pytgasync/__init__.py
Comment (parkererickson-tg) [inline pyTigerGraph/pytgasync/__init__.py:1]: This shouldn't contain version or license

**Review by parkererickson-tg — CHANGES_REQUESTED** [review]

File: tmp/Cora/._ingest.gsql
Comment (luzhoutg) [reply tmp/Cora/._ingest.gsql:1]: removed.

File: tests/testserver.json
Comment (luzhoutg) [reply tests/testserver.json:1]: fixed the typo. I modified it after the testing.

File: pyTigerGraph/pytgasync/__init__.py
Comment (luzhoutg) [reply pyTigerGraph/pytgasync/__init__.py:1]: removed

Comment (parkererickson-tg) [reply pyTigerGraph/pytgasync/__init__.py:1]: I am saying this shouldn't have __version__ or __license__ at all, should be just the from . import statement. Version and license is covered in `pyTigerGraph/__init__.py`

**Review by qe-tigergraph — COMMENTED** [review]
Unit Test: SUCCESS, e2e Test: SKIPPED, Jenkins_job:http://192.168.99.101:30080/job/mlwb_build/1150/

**Review by qe-tigergraph — APPROVED** [review]
QE Approved

**Review by parkererickson-tg — APPROVED** [review]

File: pyTigerGraph/pytgasync/schema.py [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+    else:
+        if str(attr_type).lower() in BASE_TYPES:
+            return str(attr_type).upper()
+        else:
+            raise TigerGraphException(
+                attr_type+"not a valid TigerGraph datatype.")
+
+
+@dataclass
+class Vertex(object):
```
Comment (RobRossmiller-TG) [inline pyTigerGraph/pytgasync/schema.py:178]: We don't need to redefine classes that have no changes

**Review by RobRossmiller-TG — CHANGES_REQUESTED** [review]

File: pyTigerGraph/pytgasync/pyTigerGraphDataset.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        print("---- Creating loading job ----", flush=True)
+        resp = await dataset.create_load_job(self)
+        print(resp, flush=True)
+        if "Failed" in resp:
+            return
+
+        print("---- Ingesting data ----", flush=True)
+        self.graphname = dataset.name
+        if getToken:
+            await self.getToken(self.createSecret())
```
[the remark this answers is not in the document]
Comment (luzhoutg) [reply pyTigerGraph/pytgasync/pyTigerGraphDataset.py:77]: good catch. revise to following:
`if getToken:
     secret = await self.createSecret()
      await self.getToken(secret)`

**Review by parkererickson-tg — COMMENTED** [review]
Do we need these functions to be async? These should block before running the loading jobs, as token is a required step before ingest.

**Review by qe-tigergraph — COMMENTED** [review]
Unit Test: SUCCESS, e2e Test: SKIPPED, Jenkins_job:http://192.168.99.101:30080/job/mlwb_build/1151/

**Review by qe-tigergraph — APPROVED** [review]
QE Approved

**Review by qe-tigergraph — COMMENTED** [review]
Unit Test: SUCCESS, e2e Test: SKIPPED, Jenkins_job:http://192.168.99.101:30080/job/mlwb_build/1152/

**Review by qe-tigergraph — APPROVED** [review]
QE Approved

**Comment by luzhoutg** [comment]
- cleaned up the code again to remove some unnecessary pieces.
- kicked off tests, all passed.

File: pyTigerGraph/pytgasync/pyTigerGraphAuth.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+from typing import Union
+import httpx
+
+from pyTigerGraph.pyTigerGraphException import TigerGraphException
+from pyTigerGraph.pyTigerGraphAuth import pyTigerGraphAuth
+from pyTigerGraph.pytgasync.pyTigerGraphGSQL import AsyncPyTigerGraphGSQL
+
+logger = logging.getLogger(__name__)
+
+# TODO FINISH MAKING THE ASYNCS AND ADD COMMENTS
```
Comment (RobRossmiller-TG) [inline pyTigerGraph/pytgasync/pyTigerGraphAuth.py:17]: Is this todo handled?

File: pyTigerGraph/pytgasync/pyTigerGraphAuth.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+                    res = await self._req(method, alt_url, authMode=authMode, data=alt_data, resKey=None)
+                    # res['token'] = res['results']['token'] # bringing token out to outer dict level for consistency, can't just use resKey="results" since want to preserve error message
+                    mainVer = 3
+                else:
+                    raise e
+
+        # uses mainVer instead of _versionGreaterThan4_0 since you need a token for verson checking
+        return res, mainVer
+
+    async def newGetToken(self, secret: str = None, setToken: bool = True, lifetime: int = None) -> Union[tuple, str]:
```
Comment (RobRossmiller-TG) [inline pyTigerGraph/pytgasync/pyTigerGraphAuth.py:161]: Should this be prefaced with new, or is it just getToken. Same question for refresh and delete

**Review by RobRossmiller-TG — CHANGES_REQUESTED** [review]
Check for TODO comments. Not sure if they're old or actual TODO's. Some functions are prefaced with "new" -- why?

This might be a good opportunity to stop and look at pyTG as a whole

File: pyTigerGraph/pytgasync/pyTigerGraphAuth.py (hunk 3 above)
Comment (luzhoutg) [reply pyTigerGraph/pytgasync/pyTigerGraphAuth.py:17]: This looks like old TODO's. Let me remove it.

File: pyTigerGraph/pytgasync/pyTigerGraphAuth.py (hunk 4 above)
Comment (luzhoutg) [reply pyTigerGraph/pytgasync/pyTigerGraphAuth.py:161]: I reviewed all the TODO comments in the `pytgasync` folder. It appears they were copied directly from the synchronous functions and are outdated, as they haven't been implemented in either the async or sync versions. 

Regarding functions prefixed with "new," it seems Dyllan has broken down the `getToken` function into 'newGetToken' with smaller helper functions: `_prep_newToken`, `_newToken`, and `_parse_newToken`, which can be reused in "newDeleteToken", "newRefreshToken". These new functions are used across unit tests. 

Let me try to comment out "getToken" function, and rename "newGetToken" to "getToken", and run another round test. 

I agree that we might have to reorganize and clean up pyTG as a whole. We can discuss it with the team and add it to sprint planning.

Comment (RobRossmiller-TG) [reply pyTigerGraph/pytgasync/pyTigerGraphAuth.py:161]: So the api is consistent across sync and async, they should not have the "new" prefix

**Review by qe-tigergraph — COMMENTED** [review]
Unit Test: SUCCESS, e2e Test: SKIPPED, Jenkins_job:http://192.168.99.101:30080/job/mlwb_build/1153/

**Review by qe-tigergraph — APPROVED** [review]
QE Approved

File: pyTigerGraph/pytgasync/pyTigerGraphAuth.py (hunk 4 above)
Comment (luzhoutg) [reply pyTigerGraph/pytgasync/pyTigerGraphAuth.py]: cleaned up the code by removing new prefix and duplicates, update the functions across different functions.

**Review by qe-tigergraph — COMMENTED** [review]
Unit Test: SUCCESS, e2e Test: SKIPPED, Jenkins_job:http://192.168.99.101:30080/job/mlwb_build/1154/

**Review by qe-tigergraph — APPROVED** [review]
QE Approved

**Review by qe-tigergraph — COMMENTED** [review]
Unit Test: SUCCESS, e2e Test: SKIPPED, Jenkins_job:http://192.168.99.101:30080/job/mlwb_build/1158/

**Review by qe-tigergraph — APPROVED** [review]
QE Approved

**Pull request closed without merging**